File: 프로그래머스/3/42898. 등굣길/등굣길.cpp
#include <bits/stdc++.h>
using namespace std;

int solution(int m, int n, vector<vector<int>> puddles) {
    long long dp[101][101] = {0, };
    bool isWater[101][101] = {false, };
    
    for(int i=0; i<puddles.size(); i++){
        isWater[puddles[i][1]][puddles[i][0]] = true;
    }
    
    dp[1][1] = 1;
    for(int i=1; i<=n; i++){
        for(int j=1; j<=m; j++){
            if(isWater[i][j]) dp[i][j] = 0;
            else if(i == 1 && j == 1) continue;
            else if(i == 1 || j == 1) dp[i][j] = i == 1 ? dp[i][j-1] : dp[i-1][j];
            else dp[i][j] = dp[i-1][j] % 1000000007 + dp[i][j-1] % 1000000007;
        }
    }
    
    long long answer = dp[n][m] % 1000000007;
    
    return answer;
}

File: 프로그래머스/3/42898. 등굣길/등굣길.java
class Solution {
    
    int[][] routeCount;
    boolean[][] isPuddle;
    public int solution(int m, int n, int[][] puddles) {
        
        routeCount = new int[m][n];
        isPuddle = new boolean[m][n];
        
        for (int i=0; i<puddles.length; i++) {
            isPuddle[puddles[i][0]-1][puddles[i][1]-1] = true;
        }
        
        
        routeCount[0][0] = 1;
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                if (i==0 && j==0) continue;
                if (isPuddle[i][j]) continue;
                int up = 0;
                int left = 0;
                if (i != 0) up = routeCount[i-1][j];
                if (j != 0) left = routeCount[i][j-1];
                
                routeCount[i][j] = ((up + left) % 1000000007) % 1000000007;
            }
        }
        
        int answer = routeCount[m-1][n-1];
        return answer;
    }
}